Throw Ruby Exception as RubyError in JS

Fayti1703 opened the pull request:
This replaces the previous behavior of throwing the string `"Ruby exception"` with a more JavaScript-y error throw.

Note that the Ruby exception is still stored in a single field on the Ruby isolate. This means that if multiple errors are thrown, only the first is reported.

seanmakesgames committed:
pass exception text into javascript

seanmakesgames committed:
rem debug

Fayti1703 committed:
Wrap ruby error into a JS Error class

Fayti1703 committed:
Adjust tests

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -896,6 +899,12 @@ void IsolateInfo::init(SnapshotInfo* snapshot_info) {
 
     isolate = Isolate::New(create_params);
+
+    Locker lock { isolate };
```
Can you add a comment block above here describing what is happening here? Or factor this out as a function with a comment above it describing what is happening?

Fayti1703 replied:
I could, but this doesn't really do much from a high-level point of view -- it's just creating an `ObjectTemplate` with a single internal field which we later use to create `RubyError` instances with.

(The internal field is where I originally planned to save a reference to the Ruby error instance in -- it's not accessible to JavaScript, so we can use it to smuggle another value through, such as a `v8::External`.)

EDIT: Well, turns out I already fill it with the Ruby error instance, but that's never used. So it's just a dead `void*` for now.

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -949,6 +958,10 @@ static VALUE rb_isolate_pump_message_loop(VALUE self) {
 }
 
+static void NoRubyErrorConstructor(const FunctionCallbackInfo<Value>& args) {
+    args.GetIsolate()->ThrowError("This class may not be constructed from JS!");
```
Change ! exclamation mark to a . period

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
````diff
@@ -981,8 +994,44 @@ static VALUE rb_context_init_unsafe(VALUE self, VALUE isolate, VALUE snap) {
+        /* Create `RubyError` constructor + prototype (and steal `Error.captureStackTrace` while we're at it)
+         * equivalent JS code is roughly
+         * ```
+         * function RubyError() { throw new Error("This class may not be constructed from JS!") }
````
Change ! exclamation mark to a . period

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -981,8 +994,44 @@ static VALUE rb_context_init_unsafe(VALUE self, VALUE isolate, VALUE snap) {
+
+        Local<String> strClassName = String::NewFromUtf8Literal(isolate_info->isolate, "RubyError");
+
+        /* Create `RubyError` constructor + prototype (and steal `Error.captureStackTrace` while we're at it)
```
prefer // line comments over block comment. (you can use indenting to make the block more readable.

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -949,6 +958,10 @@ static VALUE rb_isolate_pump_message_loop(VALUE self) {
 
+static void NoRubyErrorConstructor(const FunctionCallbackInfo<Value>& args) {
+    args.GetIsolate()->ThrowError("This class may not be constructed from JS!");
+}
```
ThrowError is not a function in our current version of v8. ThrowException exists, but requires a Local<> Error instance.

Fayti1703 replied:
I've inlined the definition as it appears in my version of V8.

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -981,8 +994,44 @@ static VALUE rb_context_init_unsafe(VALUE self, VALUE isolate, VALUE snap) {
+        Local<Function> captureStackTrace = errorConstructor
+            ->Get(context, String::NewFromUtf8Literal(isolate_info->isolate, "captureStackTrace"))
+            .ToLocalChecked().As<Function>();
+        context_info->capture_stack_trace.Reset(isolate_info->isolate, captureStackTrace);
```
can you explain to me using Reset here? As a brand new Local<> there shouldn't be anything to Reset from my understanding.

Fayti1703 replied:
`Global<T>::Reset(Isolate*, Local<T> value)` puts the value of `value` into the `Global<T>` it is called on.

So basically, this is just storing `captureStackTrace` in `context_info->capture_stack_trace`

seanmakesgames replied:
Understood

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -1247,8 +1322,9 @@ gvl_ruby_callback(void* data) {
             (VALUE(*)(...))&rescue_callback, (VALUE)(&callback_data), rb_eException, (VALUE)0);
 
     if(callback_data.failed) {
+        /* TODO: Use the v8 exception as a data storage instead */
```
Nit: If you are leaving a 'todo' in comments, then you should have an associated issue / work item reference. Otherwise it should change to a "NOTE: we can do a better job here if we do X."

Fayti1703 committed:
Inline Isolate::ThrowError

Fayti1703 committed:
Quieten exception message a bit

Fayti1703 committed:
Initialize the stored v8 Globals

Not doing this causes various random segmentation faults
as v8 ends up reading garbage out of uninitialized memory.

0-arg constructor on these types doesn't do anything except
initialize the internals to nullptr, so this should be fine for
the Ruby alloc method.

seanmakesgames committed:
Adjust comments

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -981,8 +994,44 @@ static VALUE rb_context_init_unsafe(VALUE self, VALUE isolate, VALUE snap) {
+        Local<Object> global = context->Global();
+        Local<Object> errorConstructor = global
+            ->Get(context, String::NewFromUtf8Literal(isolate_info->isolate, "Error"))
+            .ToLocalChecked().As<Object>();
```
Looks like all of these checks will crash ruby MRI if they fail. Which will crash the worker.
I hope they don't. But something to watch out for when we are testing if this functionality actually works. Ideally nothing we do here should cause ruby to crash.

Fayti1703 replied:
This is pretty early initialization code... plus, the existing mini_racer code was just as crash-prone.

I'd actually be surprised if any of this throws an error, though if you want that handled, that is going to be a bit more complicated, since we'd need a `v8::TryCatch` &c &c

seanmakesgames replied:
Indeed. and in general there aren't a lot of crashes at the moment. We'll know if they start showing up! :)

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -1185,6 +1234,31 @@ VALUE rescue_callback(VALUE rdata, VALUE exception) {
+	Local<Object> errorInstance;
+	Local<Context> context = context_info->context->Get(isolate);
+	if(!errorTemplate->NewInstance(context).ToLocal(&errorInstance))
+		return;
```
A few of these returns; will end up producing half-formed js error objects and not throwing an exception in JS when an error happens in Ruby.
We should probably try to throw -at least- the old 'ruby exception' we had.

Fayti1703 replied:
If those guard-`return`s are hit, the V8 internals already threw an exception and we just need to get out of the way to let V8 propagate that through JS.

Yes, the V8 API is a bit wierd like that.

seanmakesgames replied:
Ok. Sounds good.

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -1185,6 +1234,31 @@ VALUE rescue_callback(VALUE rdata, VALUE exception) {
+	if(!errorTemplate->NewInstance(context).ToLocal(&errorInstance))
+		return;
+	/* TODO: Is this cast sane? */
+	errorInstance->SetInternalField(0, External::New(isolate, (void*) rb_error));
```
Yeah, I have no idea if you can safely shove the ruby error VALUE into the field here. Seems unlikely, especially since the function is looking for a v8 Local<Value>.

Fayti1703 replied:
We wrap it into a `v8::Local<v8::External>`, so that part is fine. The TODO comment was more because I wasn't sure if we could safely cast `VALUE` into `void*` at the time.

From what I know now, that *is* a sane cast.

seanmakesgames replied:
https://github.com/ruby/ruby/blob/cb4e2cb55a59833fc4f1f6db2f3082d1ffcafc80/include/ruby/internal/value.h#L40
https://www.reddit.com/r/C_Programming/comments/dztsbu/cast_from_void_to_uintptr_t/

not 100% safe, but mostly safe. :)
Which is good enough for me!

seanmakesgames replied:
Wait-
Now that I've read and understood the code more, we don't use this data at all, right? We've already serialized this into the JS-side error RubyError and have thrown it. Are we using it somewhere I missed?

Fayti1703 replied:
The value stored here is currently unused. This is mainly prep for moving away from storing the ruby exception in `@current_exception` on the context.

seanmakesgames committed:
Update tests

seanmakesgames commented on ext/mini_racer_extension/mini_racer_extension.cc:
```diff
@@ -1513,6 +1588,8 @@ static VALUE allocate(VALUE klass) {
     ContextInfo* context_info = ALLOC(ContextInfo);
     context_info->isolate_info = NULL;
     context_info->context = NULL;
+    new (&context_info->ruby_error_prototype) Global<ObjectTemplate>();
```
What are these lines doing? This syntax is weird to me.

seanmakesgames replied:
Is this equivalent to 
`context_info->ruby_error_prototype = new Global<ObjectTemplate>();`?

Fayti1703 replied:
Not *quite*. Similar logic, but `new X()` in C++ always allocates a new `X` and returns an `X*` to it.

What I'm doing there is called "placement new" -- we already have a spot to put a `Global<ObjectTemplate>` (namely `context_info->ruby_error_prototype`), but it's uninitialized and we want to call the constructor on it.

I think in theory we could do this via `context_info->ruby_error_prototype = Global<ObjectTemplate>();` as well, but that actually does *more* stuff (that we don't need) under the hood because C++ is fun.

seanmakesgames replied:
ah yes. ye olde placement new. :)

seanmakesgames committed:
Add custom error class in test

seanmakesgames committed:
Annoyed with whitespace and fixed a bunch of them.

seanmakesgames committed:
remove platforms from ci (#6)

* remove platforms from ci

* remove stuff that didn't exist yet

* Cleaner diff

* Even less diff

seanmakesgames committed:
Error prototype is protected/frozen and failing silently
so use CreateDataProperty to do the setting.
